### DISCUSSION TITLE
automatic deskewing/dekeystoning "scanned" (photographed) sheet music (non-text)?

### DISCUSSION THREAD

**Discussion by th1rtyf0ur** [ImageMagick/ImageMagick#7815] [Help]
### ImageMagick version

7.1.1-34

### Operating system, version and so on

Mac / Linux

### Description

I have a large set of photos from my piano book collection that I took 10y ago and never got around to processing, and am trying to get through them. Lighting is fairly OK but uneven, I didn't have a tripod or other mounted setup to ensure that all pages were lined up consistently, and although I did use a glass plate to flatten the pages, some of them still have some warping near the spine (particularly in thicker books). I've already checked in the [DIY Book Scanner forums](https://www.diybookscanner.org/forum/viewtopic.php?t=3791), & they recommended checking here as well, as most of the algorithms in their software (ScanTailor, YASW, etc.) are calibrated for text. They do have a step for correcting distortion, but the auto-detection doesn't always work very well (see example pic). Since I have already dragged 4 corners (or more, if doing curved pages as well) hundreds of times and still have 2 more large volumes to go w/ nearly 600 pages between them, I'm looking for anything that can speed up the process. Each page is guaranteed to have straight-horizontal staff lines, as well as straight vertical lines along the left (w/ braces to the left of that, and a possible indent on the first set). Most pages also have vertical lines along the far right edge, unless there's a time/key signature change. Page numbers & other notes may appear in the borders. But theoretically it should be easy to detect these "straight" lines and re-straighten them, right?
Anyone have any tips on what filters/scripts I should be using (ideally the output would also be cropped to page size, but that can be done separately if necessary)? TBH I'd just re-photograph them all with a proper setup to ensure uniformity (since that would be far less effort at this point), but the physical books are in another country. 

![ScanTailor Experimental - auto-deskew detection](https://github.com/user-attachments/assets/869009d6-fe0f-459c-a420-cdd1cdb7f21a)
![original photo](https://github.com/user-attachments/assets/ac083c3d-d045-4be4-a8a3-69c35c32d553)

**Comment by snibgo** [comment]
I suggest a perspective distortion. Find four points in the image that should be the corners of a rectangle. I found those coordinates using a home-made image editor, but Gimp could be used. These coordinates start top-left, proceeding clockwise. Windows BAT syntax:
```
set x0=508
set y0=1290
set x1=3076
set y1=1352
set x2=3069
set y2=3815
set x3=449
set y3=3764
```
In principle, these four points could be found automatically, but the task is not trivial.

Simple script commands then find the coordinates of the enclosing rectangle:
```
set XL=%x0%
if %XL% GTR %x3% set XL=%x3%

set YT=%y0%
if %YT% GTR %y1% set YT=%y1%

set XR=%x1%
if %XR% LSS %x2% set XR=%x2%

set YB=%y2%
if %YB% LSS %y3% set YB=%y3%

set COORDS=%x0%,%y0% %XL%,%YT% %x1%,%y1% %XR%,%YT% %x2%,%y2% %XR%,%YB% %x3%,%y3% %XL%,%YB% 

echo %COORDS%
```
Now we can distort the image to move the original four points to the new rectangle. "presto.jpg" is my name for your image.
```
magick ^
  presto.jpg ^
  -colorspace Gray ^
  -distort Perspective "%COORDS%" ^
  -contrast-stretch 5,50%% ^
  -level 0,90%% ^
  -resize 600 -quality 40 ^
  out.jpg
```
In addition to the perspective distortion, this converts the image to gray, stretches the contrast to make 5% of the image black and 50% white, and further increases contrast to remove the bleed-through printing at bottom-right.  I haven't corrected the uneven lighting, though this is also possible.

I resize and save to low-quality jpg for internet bandwidth reasons.

We can clearly see the stave lines are curved, not straight. The curvature decreases from the top. Further processing can correct this.
![out](https://github.com/user-attachments/assets/ec3848d4-d7b9-4178-bdfb-4134dca27812)

**Comment by th1rtyf0ur** [comment]
> In principle, these four points could be found automatically, but the task is not trivial.

right- this is where I need help. I can "easily" do the perspective correction interactively in GIMP, but with nearly 600 images remaining in my queue, I'd much prefer a way to automate that part of the process. XD

**Comment by snibgo** [comment]
For the four points found manually above, I ignored the top stave, then selected the corners of the remaining staves. This can be done with connected components and morphology. This gives four points per stave. Of these, the one that is most top-left (ie x+y is the minimum) is one of the four required points. Similarly for the other three points.

With that approach, the bottom stave may be shorter than the others, so it may be better to also ignore that.

Another approach is to notice that the background is darker than the paper. The top edge has a clear boundary between the two. The bottom edge of the paper is not as clean, but it isn't bad. We can find the top edge of the paper at two places, say 20% and 80% of the width, and likewise the bottom edge of the paper. This gives us four points for the perspective distortion.

So we take a vertical slice ("-crop") through the image at 20% of the width, then moving down that slice we have: dark background, light paper, dark background. We can find those boundaries. The strip could be one pixel wide but noise would cause problems, so we use make the strips 10 pixels wide and average them ("-scale" to one pixel wide).

Windows BAT script:
```
set SRC=C:\web\im\presto.jpg

set Thrsh=80

for /F "usebackq" %%L in (`%IMG7%magick ^
  %SRC% ^
  -format "WW=%%w\nHH=%%h\nx0=%%[fx:floor(w*0.2+0.5)]\nx1=%%[fx:floor(w*0.8+0.5)]\n"
  info:`) do set %%L

set x3=%x0%
set x2=%x1%

echo WW=%WW% HH=%HH% x0=%x0% x1=%x1% x2=%x2% x3=%x3%

for /F "usebackq tokens=1-4 delims=x+" %%A in (`%IMG7%magick ^
  %SRC% ^
  -colorspace Gray ^
  -contrast-stretch "5,50%%" ^
  -write tmp.miff ^
  -crop 10x%%h+%x0%+0 +repage ^
  -scale "1x%%h^!" ^
  -write x0.png ^
  -threshold %Thrsh%%% ^
  -bordercolor Black -border 1 ^
  -format %%@ ^
  info:`) do (
  echo %%A %%B %%C %%D
  set /A y0=%%D-1
  set /A y3=%%B+%%D-1
)

echo y0=%y0%  y3=%y3%

for /F "usebackq tokens=1-4 delims=x+" %%A in (`%IMG7%magick ^
  tmp.miff
  -crop 10x%%h+%x1%+0 +repage ^
  -scale "1x%%h^!" ^
  -write x1.png ^
  -threshold %Thrsh%%% ^
  -bordercolor Black -border 1 ^
  -format %%@ ^
  info:`) do (
  echo %%A %%B %%C %%D
  set /A y1=%%D-1
  set /A y2=%%B+%%D-1
)

echo y1=%y1%  y2=%y2%

rem Make a debugging image that shows the four points.
%IMG7%magick ^
  %SRC% ^
  -fill sRGBA(0,100%%,100%%,0.5) ^
  -draw "translate %x0%,%y0% circle 0,0 0,50" ^
  -draw "translate %x1%,%y1% circle 0,0 0,50" ^
  -draw "translate %x2%,%y2% circle 0,0 0,50" ^
  -draw "translate %x3%,%y3% circle 0,0 0,50" ^
  -resize 600 -quality 40 ^
  dbg.jpg

set XL=%x0%
if %XL% GTR %x3% set XL=%x3%

set YT=%y0%
if %YT% GTR %y1% set YT=%y1%

set XR=%x1%
if %XR% LSS %x2% set XR=%x2%

set YB=%y2%
if %YB% LSS %y3% set YB=%y3%

set COORDS=%x0%,%y0% %XL%,%YT% %x1%,%y1% %XR%,%YT% %x2%,%y2% %XR%,%YB% %x3%,%y3% %XL%,%YB% 

echo %COORDS%

%IMG7%magick ^
  %SRC% ^
  -colorspace Gray ^
  -distort Perspective "%COORDS%" ^
  -contrast-stretch 5,50%% ^
  -level 0,90%% ^
  -resize 600 -quality 40 ^
  out.jpg
```

As always, lines like "-write x?.png" are for debugging.

The script also makes a debugging image that shows the four found points:

![dbg](https://github.com/user-attachments/assets/6b72227d-028a-4a54-953d-442a6f505155)

We can see the found point at bottom-left isn't ideal.

The result, out.jpg, is:

![out](https://github.com/user-attachments/assets/0686be32-b32e-4de7-9be1-b4a300068934)
